File: scala/bob/src/main/scala/Bob.scala
class Bob {
  def hey(input: String): String = input match {
    case IsShouting() => "Whoa, chill out!"
    case IsQuestion() => "Sure."
    case IsSilence() => "Fine. Be that way!"
    case _ => "Whatever."
  }

  case object IsSilence {
    def unapply(input: String) = input.trim.isEmpty
  }

  case object IsQuestion {
    def unapply(input: String) = input.endsWith("?")
  }

  case object IsShouting {
    def unapply(input: String) =
      input == input.toUpperCase && input.matches(".*[A-Z].*")
  }
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
        is_silence?(input)  -> "Fine. Be that way!"
        is_question?(input) -> "Sure."
        is_shouting?(input) -> "Whoa, chill out!"
        true                -> "Whatever."
    end
  end

  defp is_silence?(input) do
    String.trim(input) == ""
  end

  defp is_shouting?(input) do
    input == String.upcase(input) && Regex.match?(~r/\p{L}+/, input)
  end

  defp is_question?(input) do
    String.ends_with?(input, "?")
  end
end
